### PR TITLE
pmi: warning fixes and interface cleanup

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

* correct MPIR_pmi_kvs_put to use PMIX_GLOBAL
* add `const` to MPIR_pmi_kvs_put/get parameters
* split MPIR_pmi_barrier into generic and local versions

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Passes tests (included warning check)
* [x] Confirm whitespace/style checkers are happy (or has a good reason for being bad)
* [x] Commits are self-contained and do not do two things at once
* [ ] Remove xfail from the test suite when fixing a test
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/warnings/more
test:jenkins/ch4/most
test:jenkins/ch3/most

**Comment by hzhou** [comment]
@hajimefu Would you like to review this?

File: src/util/mpir_pmi.c [hunk 1]
```diff
@@ -264,42 +264,51 @@ int MPIR_pmi_barrier(MPIR_PMI_DOMAIN domain)
...
+                         "**pmix_fence %d", pmi_errno);
+
+    PMIX_VALUE_RELEASE(pvalue);
+    PMIX_PROC_FREE(procs, nprocs);
 
   fn_exit:
     return mpi_errno;
   fn_fail:
     goto fn_exit;
+#else
```
Comment (hajimefu) [inline src/util/mpir_pmi.c]: Probably somewhat helpful to put a comment on what this `else` is for. `!USE_PMIX_API` case, I guess?

File: src/include/mpir_pmi.h [hunk 2]
```diff
@@ -45,12 +45,14 @@ const char *MPIR_pmi_job_id(void);
...
 
-/* * barrier or kvs fence. "domain" is a hint for efficiency (eg PMIx) */
-int MPIR_pmi_barrier(MPIR_PMI_DOMAIN domain);
+/* * barrier or kvs fence. */
+int MPIR_pmi_barrier(void);
+/* * barrier over local set. More efficient for PMIx. Same as MPIR_pmi_barrier for PMI1/2. */
+int MPIR_pmi_barrier_local(void);
 /* * put, to global domain */
-int MPIR_pmi_kvs_put(char *key, char *val);
+int MPIR_pmi_kvs_put(const char *key, const char *val);
```
Comment (hajimefu) [inline src/include/mpir_pmi.h]: To me this commit seems to mix two things in one (barrier change and const). I would split it into two patches.

Comment (hzhou) [reply src/include/mpir_pmi.h]: Thanks for the review. I'll make the change.

**Comment by hzhou** [comment]
test:mpich/warnings
test:jenkins/ch3/sock
test:jenkins/ch4/most

**Comment by hzhou** [comment]
@hajimefu I have addressed your comments. Could you take another look?

**Pull request merged**